plugins/woocommerce/includes/class-wc-cart-session.php

```diff
@@ -188,16 +187,40 @@ public function get_cart_from_session() {
 				 */
 				$message = apply_filters( 'woocommerce_cart_item_removed_message', $message, $product );
 				wc_add_notice( $message, 'error' );
+
+				/**
+				 * Fires when cart item is removed from the session.
+				 *
+				 * @since 3.6.0
+				 *
+				 * @param string     $key Cart item key.
+				 * @param array      $values Cart item values e.g. quantity and product_id.
+				 */
 				do_action( 'woocommerce_remove_cart_item_from_session', $key, $values );
 
 			} elseif ( ! empty( $values['data_hash'] ) && ! hash_equals( $values['data_hash'], wc_get_cart_item_data_hash( $product ) ) ) { // phpcs:ignore PHPCompatibility.PHP.NewFunctions.hash_equalsFound
 				$update_cart_session = true;
 				/* translators: %1$s: product name. %2$s product permalink */
 				$message = sprintf( __( '%1$s has been removed from your cart because it has since been modified. You can add it back to your cart <a href="%2$s">here</a>.', 'woocommerce' ), $product->get_name(), $product->get_permalink() );
-				$message = apply_filters( 'woocommerce_cart_item_removed_because_modified_message', $message, $product );
-				wc_add_notice( $message, 'notice' );
-				do_action( 'woocommerce_remove_cart_item_from_session', $key, $values );
 
+				/**
+				 * Filter message about item removed from the cart because it has since been modified.
+				 *
+				 * @since 3.8.0
+				 * @param string     $message Message.
+				 * @param WC_Product $product Product data.
+				 */
+				wc_add_notice( apply_filters( 'woocommerce_cart_item_removed_because_modified_message', $message, $product ), 'notice' );
+
+				/**
+				 * Fires when cart item is removed from the session.
+				 *
+				 * @since 3.6.0
+				 *
+				 * @param string     $key Cart item key.
+				 * @param array      $values Cart item values e.g. quantity and product_id.
+				 */
+				do_action( 'woocommerce_remove_cart_item_from_session', $key, $values );
```

Comment:
I know it already exists but should this be done here to begin with? why not at normalize stage later down the road.